bdk_core/src/coin_select.rs

```diff
@@ -1,77 +1,128 @@
 
-const TXIN_BASE_WEIGHT: u32 = (32 + 4 + 4) * 4;
+pub const TXIN_FIXED_WEIGHT: u32 = (32 + 4 + 4) * 4;
 
 #[derive(Debug, Clone)]
-pub struct CoinSelector {
-    candidates: Vec<WeightedValue>,
+pub struct CoinSelector<'a> {
+    candidates: Vec<InputCandidate>,
     selected: BTreeSet<usize>,
-    opts: CoinSelectorOpt,
+    opts: &'a CoinSelectorOpt,
 }
 
 #[derive(Debug, Clone, Copy)]
-pub struct WeightedValue {
+pub struct InputCandidate {
     pub value: u64,
+    /// Weight of the `txin`: `prevout` + `nSequence` + `scriptSig` + `scriptWitness`.
     pub weight: u32,
-    pub is_segwit: bool,
+    /// Number of inputs contained within this [`InputCandidate`].
+    /// If we are using single UTXOs as candidates, this would be 1.
+    /// If we are working in `OutputGroup`s (as done in Bitcoin Core), this would be > 1.
+    pub input_count: usize,
+    /// Whether this `txin` is spending a segwit output.
+    pub has_segwit: bool,
 }
 
-#[derive(Debug, Clone, Copy)]
+#[cfg(feature = "std")]
+impl std::ops::Add for InputCandidate {
+    type Output = Self;
+
+    fn add(self, other: Self) -> Self {
+        Self {
+            value: self.value + other.value,
+            weight: self.weight + other.weight,
+            input_count: self.input_count + other.input_count,
+            has_segwit: self.has_segwit || other.has_segwit,
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
 pub struct CoinSelectorOpt {
-    /// The value we need to select.
+    /// The value we need to select (in satoshis).
     pub target_value: u64,
     /// The feerate we should try and achieve in sats per weight unit.
     pub target_feerate: f32,
-    /// The minimum absolute fee.
+    /// The long term feerate if we are to spend an input in the future instead (in sats/wu).
+    /// This is used for calculating waste.
+    pub long_term_feerate: f32,
+    /// The minimum absolute fee (in satoshis).
     pub min_absolute_fee: u64,
-    /// The weight of the template transaction including fixed inputs and outputs.
-    pub base_weight: u32,
-    /// The weight of the drain (change) output.
+
+    /// Additional weight if we use the drain (change) output(s).
     pub drain_weight: u32,
-    /// The input value of the template transaction.
-    pub starting_input_value: u64,
+    /// Weight of a `txin` used to spend the drain output(s) later on.
+    pub drain_spend_weight: u32,
+
+    /// The fixed input(s) of the template transaction (if any).
+    pub fixed_input: Option<InputCandidate>,
+    /// The additional fixed weight of the template transaction including: `nVersion`, `nLockTime`
+    /// and fixed `vout`s and the first bytes of `vin_len` and `vout_len`. Weight of `vin` is not
+    /// included.
+    pub fixed_additional_weight: u32,
 }
 
 impl CoinSelectorOpt {
-    pub fn from_weights(base_weight: u32, drain_weight: u32) -> Self {
+    pub fn from_weights(fixed_weight: u32, drain_weight: u32, drain_spend_weight: u32) -> Self {
```

Comment:
base seems to be more appropriate since its the starting weight that you add to.

Reply:
I was getting it confused with "base size" as in [BIP-0141](https://github.com/bitcoin/bips/blob/master/bip-0141.mediawiki#block-size). How about `starting_weight`?